apps/lib/apps.c

```diff
@@ -2370,41 +2366,60 @@ int raw_write_stdout(const void *buf, int siz)
  * does impact behavior on some platforms, such as differentiating between
  * text and binary input/output on non-Unix platforms
  */
+static BIO *check_bio(BIO *bp, const char *what, int mode)
```

Comment:
it can't be a char, because it's an bitmask (e.g., "BIO_F_ NOCLOSE|" is often or'd in).

Reply:
Hmm, I guess you meant `BIO_NOCLOSE`, which happens to be `0x00`,
yet also `BIO_FP_TEXT` is often or'ed in, which is non-zero,
to be used as argument to
```
BIO *BIO_new_fp(FILE *stream, int flags);
```

I meanwhile found that in  `check_bio` you are mixing those BIO flags not only with the `fopen()` mode,
having values like `'r'`, but also with  `O_WRONLY`, optionally or'ed with `O_BINARY` etc.,
which is pretty unclean.

I recommend changing the function at least to
```
static BIO *check_bio(BIO *bp, const char *what, char mode, int flags);
```
to avoid confusion and for properly printing the fopen mode (if non-NUL).

Reply:
An alternative would be to rename the parameter to the less confusing
```
static BIO *check_bio(BIO *bp, const char *what, int flags);
```
and make sure that the function is called only with some flavor of flags but not with the fopen mode (such as `'r'`).

Reply:
okay, done.  additional commit pushed.

Reply:
Thanks, but the confusing/wrong name `mode` still appears in the error message
and as local var name in `dup_bio_*()` and `app_load_config_quiet()`.

Reply:
It's called "format" throughout the apps.  Changing it here would break the larger consistency.

Reply:
commit pushed to fix the error message.

Reply:
> It's called "format" throughout the apps. Changing it here would break the larger consistency.

Not really - did you mean "mode"?

There are many correct uses of the names "mode" and "format" in (hitherto) apps.c:
```
static const char *modestr(char mode, int format)
static const char *modeverb(char mode)
static BIO *bio_open_default_(const char *filename, char mode, int format,
                              int quiet)
BIO *bio_open_default(const char *filename, char mode, int format)
BIO *bio_open_default_quiet(const char *filename, char mode, int format)

```

So far there has been only one function in apps.c that confusingly has been using the name "mode" where "flags" would have been appropriate:
```
BIO *bio_open_owner(const char *filename, int format, int private)
{
...
    mode = O_WRONLY;
#ifdef O_CREAT
    mode |= O_CREAT;
#endif
#ifdef O_TRUNC
    mode |= O_TRUNC;
#endif
    textmode = FMT_istext(format);
    if (!textmode) {
#ifdef O_BINARY
        mode |= O_BINARY;
#elif defined(_O_BINARY)
        mode |= _O_BINARY;
#endif
    }
...
        fd = open(filename, mode, 0600);

```
Clearly `mode` should better have been named `flags` in this function, 
not only to avoid confusion with `char mode` used elsewhere in the file,
but also for consistency with `int open(const char *pathname, int flags);`
See also, e.g., https://linux.die.net/man/3/open

Reply:
I renamed mode to flags.

And I meant "format" in my comment above.  As in ```-outform DER``` and so on.

Reply:
Nice. The same would be good also for 
- `int mode = BIO_FP_TEXT;`
- `static BIO *check_bio(BIO *bp, const char *what, int mode)`
- `int mode = BIO_NOCLOSE | (FMT_istext(format) ? BIO_FP_TEXT : 0);`  (3 times)

Reply:
See comment below about mode/format confusion and my unwillingness to fix it in this PR.